teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,702 @@
+kind: PipelineRun
+metadata:
+  generateName: platforms-with-results
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: matrix-emitting-results
+        matrix:
+          params:
+            - name: platform
+              value:
+                - linux
+                - mac
+                - windows
+            - name: browser
+              value:
+                - chrome
+                - safari
+                - firefox
+        taskSpec:
+          params:
+            - name: platform
+              default: ""
+            - name: browser
+              default: ""
+          results:
+            - name: report-url
+              type: string
+          steps:
+            - name: produce-report-url
+              image: alpine
+              script: |
+                echo "Running tests on $(params.platform)-$(params.browser)"
+                echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+      - name: task-consuming-results
+        taskRef:
+          name: stringtask
+          kind: Task
+        params:
+          - name: url
+            Value: $(tasks.matrix-emitting-results.results.report-url[*])
+        taskSpec:
+          params:
+            - name: url
+              type: string
+          steps:
+            - name: echo
+              image: alpine
+              script: |
+                for arg in "$@"; do
+                  echo "Arg: $arg"
+                done
+```
+
+In the example below, a user wants to run tests on a combination of different platforms and browsers and then fetch the reports for only linux platforms.
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: platforms-with-results
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: matrix-emitting-results
+        matrix:
+          params:
+            - name: platform
+              value:
+                - linux
+                - mac
+                - windows
+            - name: browser
+              value:
+                - chrome
+                - safari
+                - firefox
+        taskSpec:
+          params:
+            - name: platform
+              default: ""
+            - name: browser
+              default: ""
+          results:
+            - name: report-url
+              type: string
+          steps:
+            - name: produce-report-url
+              image: alpine
+              script: |
+                echo "Running tests on $(params.platform)-$(params.browser)"
+                echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+      - name: task-consuming-results
+        taskRef:
+          name: stringtask
+          kind: Task
+        params:
+          - name: url
+            value:
+              - $(tasks.matrix-emitting-results.results.report-url[0]) # linux-chrome
```

Comment:
spitballing here, if the use case is filter all the results where `params.platform == linux`, we could add some new syntax similar to JSONPath filters e.g. `$(tasks.matrix-emitting-results.results.report-url|?(@params.platform.value=='linux')]`